### ISSUE TITLE
[GR-38912] Valgrind reports "Possibly corrupted debuginfo file" for `libjsvm.so` from Graal.js.

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#4598]
Here's an issue for @adinn (can't assign the issue to you, working on fixing that).

#### Steps to reproduce
```bash
mkdir tmp && cd tmp
git clone https://github.com/oracle/graal
git clone https://github.com/oracle/graaljs
cd graal/vm
mx --dy /graal-js,/substratevm --extra-image-builder-argument=-g --native-images=lib:jsvm build
export GRAALVM_HOME="$(mx --dy /graal-js,/substratevm --extra-image-builder-argument=-g --native-images=lib:jsvm graalvm-home)"
valgrind --tool=callgrind $GRAALVM_HOME/bin/js --eval "console.log('hello')"
```
#### Error
```
==322765== Callgrind, a call-graph generating cache profiler
==322765== Copyright (C) 2002-2017, and GNU GPL'd, by Josef Weidendorfer et al.
==322765== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==322765== Command: /home/fniephaus/dev/graalvm/graal/sdk/mxbuild/linux-amd64/GRAALVM_3ADDE48CEF_JAVA11/graalvm-3adde48cef-java11-22.2.0-dev/bin/js --eval console.log('hello')
==322765== 
==322765== For interactive control, run 'callgrind_control -h'.
==322765== Valgrind: debuginfo reader: ensure_valid failed:
==322765== Valgrind:   during call to ML_(img_get)
==322765== Valgrind:   request for range [396742070, +4) exceeds
==322765== Valgrind:   valid image size of 252542168 for image:
==322765== Valgrind:   "/home/fniephaus/dev/graalvm/graal/sdk/mxbuild/linux-amd64/GRAALVM_3ADDE48CEF_JAVA11/graalvm-3adde48cef-java11-22.2.0-dev/languages/js/lib/libjsvm.so"
==322765== 
==322765== Valgrind: debuginfo reader: Possibly corrupted debuginfo file.
==322765== Valgrind: I can't recover.  Giving up.  Sorry.
==322765==
```

Any idea what could be going on here?

**Comment by adinn** [comment]
@fniephaus I have built an image and run into the same problem. I know what is wrong but am not sure exactly why it is happening.

My analysis of the image and the debug info it contains shows that a single byte has been dropped out of the debug info section:

The header info for image includes 

     27 .debug_info   00abfdfc  0000000000000000  0000000000000000  0a8aca38  2**3

i.e. debug info section is at offset  `0xa8aca38` in the image file

The actual error in the debug info section is in the CU which begins at offset:

- `0xb2823df` in the image
- `0x9d59a7` in the .debug_info section

Here is the relevant hex dump:

    0b2823d0: 0100 1008 5459 9d00 1408 8059 9d00 00(78  ....TY.....Y...x
    0b2823e0: 0000 00)(04 00)(00 0000 00)(08) (02)(0b) (1f)(ef 5a00  ..............Z.
    0b2823f0: e7)(18 5f00 e0)(c5 5103 0000 0000 88)(c7 5103  .._...Q.......Q.
    0b282400: 0000 0000 92)(d5 0900 0c)(e0 c551 03)00 0000  ...........Q....
    0b282410: 00a5 c651 0300 0000 0001 ba07 1e00 27d1  ...Q..........'.
    0b282420: 071e 00dd c3d7 0127 dc07 1e00 3bc3 d701  .......'....;...

(note that I have bracketed successive DWARF fields)

This gets decompiled by objdump as follows

    Compilation Unit @ offset 0x9d59a7:
       Length:        0x78 (32-bit)
       Version:       4
       Abbrev Offset: 0x0
       Pointer Size:  8
     <0><9d59b2>: Abbrev Number: 2 (DW_TAG_compile_unit)
        <9d59b3>   DW_AT_language    : 11	(Java)
        <9d59b4>   DW_AT_use_UTF8    : 31
        <9d59b5>   DW_AT_name        : (indirect string, offset: 0xe7005aef): <offset is too big>
        <9d59b9>   DW_AT_comp_dir    : (indirect string, offset: 0xe0005f18): <offset is too big>
        <9d59bd>   DW_AT_low_pc      : 0x88000000000351c5
        <9d59c5>   DW_AT_high_pc     : 0x92000000000351c7
        <9d59cd>   DW_AT_stmt_list   : 0xc0009d5
     <1><9d59d1>: Abbrev Number: 1336032 !!! blows up here !!!

The error is a missing 01 byte for attribute DW_AT_use_UTF8 at offset
0x9d59b4 (address in image 0xb2823ec).

    0b2823d0: 0100 1008 5459 9d00 1408 8059 9d00 00(78  ....TY.....Y...x
    0b2823e0: 0000 00)(04 00)(00 0000 00)(08) (02)(0b) (*01*) (1fef 5a00)  ..............Z.
                                                        ^^^^
    0b2823f0: (e718 5f00) (e0c5 5103 0000 0000) (88c7 5103  .._...Q.......Q.
    0b282400: 0000 0000) (92d5 0900) (0c)(e0 c551 0300 0000  ...........Q....
    0b282410: 00)(a5 c651 0300 0000 00)(01) (ba07 1e00) 27d1  ...Q..........'.
    0b282420: 071e 00dd c3d7 0127 dc07 1e00 3bc3 d701  .......'....;...

If we adjust for that missing byte we get the following parse (follow
the brackets above).

    Compilation Unit @ offset 0x9d59a7:
       Length:        0x78 (32-bit)
       Version:       4
       Abbrev Offset: 0x0
       Pointer Size:  8
     <0><9d59b2>: Abbrev Number: 2 (DW_TAG_compile_unit)
        <9d59b3>   DW_AT_language    : 11	(Java)
        <9d59b4>   DW_AT_use_UTF8    : 01 
        <9d59b5>   DW_AT_name        : (indirect string, offset: 0x5aef1f): Module.java
        <9d59b9>   DW_AT_comp_dir    : (indirect string, offset: 0x005f18e7): /home/adinn/tmp/debuginfo/issues/4598/graal/sdk/mxbuild/linux-amd64/libjsvm.so.image/sources
        <9d59bd>   DW_AT_low_pc      : 0x351c5e0
        <9d59c5>   DW_AT_high_pc     : 0x351c788
        <9d59cd>   DW_AT_stmt_list   : 0x9d592
     <1><9d59d1>: Abbrev Number: 12 (DW_TAG_subprogram)    [has children)
        <9d59d2>   DW_AT_low_pc      : 0x351c5e0
        <9d59da>   DW_AT_high_pc     : 0x351c6a5
        <9d59e2>   DW_AT_external    : 1
        <9d59e3>   DW_AT_specification: <0x1e07ba> (method Module.ModuleRequest())

which all makes sense.

So, I am not clear what has deleted the 01 byte. However, I'll note that there are many preceding CUs with Abbrev Number 2 which have do include this byte. So, I don't think this is likely to be a simple bug in the generator itself. I'll rerun the build with detailed trace enabled and see what the generator code thinks it is printing into the info section. I'll let you know what I find.

**Comment by adinn** [comment]
@fniephaus Hi Fabio. I found the problem as soon as I obtained a trace file for the generator output. It *is* just a simple generator bug.

The byte is missing because the generator fails to write the DW_AT_use_UTF8 attribute when writing a class unit for a class that contains deopt methods. Any such class has to have two CUs because the standard and deopt methods occupy different ranges. The  use_UTF8attribute was added to the abbrev record when we switched to using UTF8 text. I added a corresponding attribute write to the generator code for the CU that declares and defines the non-deopt methods but I failed to update the corresponding generator code for the CU that declares and defines the deopt variants. I have a short (2 line) patch which should fix the problem.

I still need to test it fixes the problem but i am sure this is the error. I'll post a PR as soon as I have confirmed it works.

**Comment by adinn** [comment]
@fniephaus @olpaw By the way, what this issue makes clear is that the debuginfo gate test does not exercise the case where the generated image includes deopt methods.

Note by that I don't mean to imply that we have to ensure the test actually runs the deopt methods, never mind anything more complicated like breaking and checking them in the debugger. That would be nice, but we have a more basic problem here.

What is needed as a first step is to ensure that the generator code for deopt methods gets exercised during the test. It is clear this cannot be happening because I regularly use objdump to dump info for the gate test image and I have not seen it fail as per this issue.

Is there some way we can easily address this gap in the test coverage?

**Comment by adinn** [comment]
n.b. fix now available via PR #4605

**Comment by zakkak** [comment]
Fix merged via https://github.com/oracle/graal/commit/a26cea5a502fdf09d4bd0c2b31351d020904c771
